_posts/success_stories/2024-05-20-lavinmq-understanding-code.md

```diff
@@ -0,0 +1,50 @@
+
+LavinMQ includes all the features people want the most and leaves just a small memory footprint, supporting:
+
+* Streaming
+* Flexible Routing
+* AMQP 0.9.1 Protocol
+* Acknowledgments and Confirms
+* Stream Queue
+* Unlimited Queue Length (dependent on disk space)
+* Replication
+
+We highlighted the characteristics that stand out from a technical point of view. How does it manage to handle more than a million messages per second, and have a low memory footprint? Half of it is due to, and we quote:
+
+> Messages are written to the disk really fast and the caching is managed by the OS itself. This leads to a performance boost while keeping RAM footprint low.
+
+But the other half, as mentioned in [their blog](https://lavinmq.com/blog/crystal-clear-message-brokering), is due to it being developed in Crystal. And this comes in several interesting ways.
+
+1. Crystal is efficiently compiled thanks to its LLVM backend. This enables aggressive optimizations, like inlining methods and removing indirections.
+
+2. Crystal helps developers allocate their data either on the heap or on the stack. Heap allocations are expensive, so allocating data on the stack in the hot-path of the program brings an important boost in performance. Crystal makes it easy to choose where data should be stored.
+
+3. Crystal's stdlib and compiler are written in Crystal itself. One of Crystal's golden characteristics is its readability, making it possible to read the code that is going to be run by the program without having to resort to another, typically lower-level  language.
+
+The latter is the main point of this post. In fact, the documentation of the standard library points to the code, making the performance implications of calling a certain method explicit. As an example that will come up later, the documentation for the `file` method of the `Digest` class can be found [here](https://crystal-lang.org/api/1.12.1/Digest.html#file%28file_name%3APath%7CString%29%3Aself-instance-method). Clicking on [[View source](https://github.com/crystal-lang/crystal/blob/4cea10199/src/digest/digest.cr#L214)] we land in the method’s code.
+
+The importance of being able to read, understand, and even contribute to the language and its stdlib is exemplified in [one](https://github.com/crystal-lang/crystal/pull/13780) of the [several PRs](https://github.com/crystal-lang/crystal/pulls?q=is%3Apr+author%3Acarlhoerberg+) that 84codes contributed to Crystal. Details are not important, but if you’re curious, they noted that in the aforementioned `Digest` class’s method there was a duplication of buffering. The following picture shows the contribution: you can find in green the added line and comment explaining why it makes sense to remove the buffering in the `io` object.
+
+![diff](/assets/blog/2024-05-20-digest-diff.png)

```

Comment:
**suggestion:** Display the diff directly. That's more accessible and readable than a picture.

````suggestion
```diff
diff --git a/src/digest/digest.cr b/src/digest/digest.cr
index e6a401e90545..bc38599da7f8 100644
--- a/src/digest/digest.cr
+++ b/src/digest/digest.cr
@@ -213,6 +213,7 @@ abstract class Digest
   # Reads the file's content and updates the digest with it.
   def file(file_name : Path | String) : self
     File.open(file_name) do |io|
+      # `#update` works with big buffers so there's no need for additional read buffering in the file
+      io.read_buffering = false
       self << io
     end
   end
```
````

This needs updates to the reference above (`s/picture/diff/`) and remove of the image file.